js_hw_11/task_02.ts

```diff
@@ -0,0 +1,81 @@
+// Создайте дженерик класс Storage<T>, где T должен быть ограничен объектом, имеющим КАК МИНИМУМ {id: number}.
+// Задача класса - хранить объекты типа Т в приватном массиве
+// Реализуйте в классе следующие методы:
+//   - constructor должен принимать необзятельный массив объектов соответствующего типа. 
+//     Если массив пришел - объекты запушить в хранилище.
+//   - add, принимающий либо объект типа Т, либо объект типа Т без id. Метод должен быть реализовать с помощью ПЕРЕГРУЗКИ.
+//     Если на вход подан объект без айди - айди надо сгенерировать, затем запушить обьект в хранилище
+//     Если на вход подан объект с айди - запушить его в хранилище
+//     Для типизации используйте Utility Types
+//   - update, принимающий объект с айди и любым набором остальных ключей из типа Т. 
+//   - remove, принимающий на вход id и удаляющий объект из массива
+//   - getById(id), возвращающий объект по айди если найден
+//   - getAll(), возвращает все объекты в хранилище
+
+type TObj = {
+    id: number
+};
+
+class MyStorage<T extends TObj>{
+    private data: T[] = [];
+    constructor(arr?: T[]){
+        if (arr) {
+            this.data.push(...arr)
+        }
+    };
+    private getNewId(objArr:T[]):number {
+        let newId = 0;
+        objArr.forEach((obj:T) => {
+            if (newId < obj.id) newId = obj.id            
+        });
+        return newId + 1;
+    };
+
+    add(obj: T): void;
+    add(obj: Omit<T, "id">): void;
+    add(obj: T | Omit<T, "id">){
+        if(!("id" in obj)){
+            const id = this.getNewId(this.data);
+            const newObj = {...obj, id}
+            this.data.push(newObj as T)
+        } else {            
+            this.data.push(obj)
+        }
+    };
+    update(obj: Pick<T, "id"> & Partial<T>):void{
+        const index = this.data.findIndex((item) => item.id == obj.id);
```

Comment:
Move this find index to private method, cause it is used at least in 2 methods

Reply:
Done